src/hotspot/cpu/arm/icache_arm.hpp

```diff
@@ -32,8 +32,8 @@ class ICache : public AbstractICache {
  public:
   enum {
     stub_size      = 32,                // Size of the icache flush stub in bytes
-    line_size      = BytesPerWord,      // conservative
-    log2_line_size = LogBytesPerWord    // log2(line_size)
+    log2_line_size = 6,
```

Comment:
The value of log2_line_size is too low for FreeBSD; `LogBytesPerWord = 2` and `BytesPerWord  = 1 << LogBytesPerWord;` An instruction cache line lenght is either 32 bytes or 64 bytes (it depends of the cpu). With such a low value, the end result is a SIGILL in a perfectly supported instruction (sometime it's a sigsegv)

[LogBytesPerWord](https://github.com/battleblow/openjdk-jdk11u/blob/master/src/hotspot/share/utilities/globalDefinitions.hpp#L120)
[BytesPerWord](https://github.com/battleblow/openjdk-jdk11u/blob/master/src/hotspot/share/utilities/globalDefinitions.hpp#L126)

Reply:
It works on cortex-a7, cortex-a8, cortex-a9 and cortex-a15 with `log2_line_size = 6`

Should I patch src/hotspot/share/utilities/globalDefinitions.hpp instead?

Reply:
Does 8 work instead of 6?  I'm curious as to how big this has to be and how you arrived at 6.

Beyond that, my next question would be whether there is a value that would work on both Linux and FreeBSD?  If there is, then we can potentially fix up a define in this file for the line size that makes sense for both.  If not then the data should come a platform specific header.

Reply:
Yes it works with 8. I put 6 because the icache line length of my cpu is 64 (2**6) and 5 was causing sigill/sigsegv. This value is used for cache maintenance with __clear_cache, putting too high value can cause sigsegv.
I don't know linux at all and I don't know why it works for them :(

Reply:
I'll suggest something like this then

```
#ifdef __linux__
#define LINE_SIZE BytesPerWord
#define LOG_LINE_SIZE LogBytesPerWord
#elif _ALLBSD_SOURCE
#define LINE_SIZE BytesPerLong
#define LOG_LINE_SIZE LogBytesPerLong
#else 
#error "Platform not supported"
#endif
```

And then here use LINE_SIZE and LOG_LINE_SIZE instead of directly using either values from globals or hardcoded values.

Reply:
BytesPerLong / LogBytesPerLong are still too small (I get a line_size of 8).
How about this instead:
```
--- icache_arm.hpp.orig 2019-09-02 16:55:14.234650000 +0200
+++ icache_arm.hpp      2019-09-02 16:54:57.852151000 +0200
@@ -25,6 +25,10 @@
 #ifndef CPU_ARM_VM_ICACHE_ARM_HPP
 #define CPU_ARM_VM_ICACHE_ARM_HPP
 
+#if defined(__FreeBSD__)
+#include <machine/param.h>
+#endif
+
 // Interface for updating the instruction cache.  Whenever the VM modifies
 // code, part of the processor instruction cache potentially has to be flushed.
 
@@ -32,8 +36,13 @@ class ICache : public AbstractICache {
  public:
   enum {
     stub_size      = 32,                // Size of the icache flush stub in bytes
+#if defined(__FreeBSD__)
+    log2_line_size = CACHE_LINE_SHIFT,
+    line_size      = 1 << log2_line_size
+#else
     line_size      = BytesPerWord,      // conservative
     log2_line_size = LogBytesPerWord    // log2(line_size)
+#endif
   };
 };
```
CACHE_LINE_SHIFT is defined [here](https://svnweb.freebsd.org/base/head/sys/arm/include/param.h?revision=325966&view=markup#l112)

Reply:
I'm fine with something like that

Reply:
This is bug in FreeBSD kernel. I just committed fix  for this issue , which will be MFCed to 13.0, 13-stable and 12-stable.  - https://cgit.FreeBSD.org/src/commit/?id=ec090f4a67654fa541e6d97fd5f74d3f66c1c0d0
For 12.1 and 12.2 we can hold local patch in port tree,  I think.